Sending telegram message accomplish with ConcurrentModificationException

When server sending list of cars. occurs that exception.

```
2022-08-22 01:49:01.012 ERROR 412009 --- [nder-thread-353] reactor.core.publisher.Operators         : Operator called default onErrorDropped

reactor.core.Exceptions$ErrorCallbackNotImplemented: java.util.ConcurrentModificationException
Caused by: java.util.ConcurrentModificationException: null
	at java.base/java.util.ArrayList$Itr.checkForComodification(ArrayList.java:1013) ~[na:na]
	at java.base/java.util.ArrayList$Itr.next(ArrayList.java:967) ~[na:na]
	at by.miaskor.bot.service.cf4j.LinkPropertiesProvider.getProperties(LinkPropertiesProvider.kt:115) ~[main/:na]
	at org.cfg4j.source.files.FilesConfigurationSource.getConfiguration(FilesConfigurationSource.java:127) ~[cfg4j-core-4.4.1.jar:na]
	at org.cfg4j.provider.SimpleConfigurationProvider.getProperty(SimpleConfigurationProvider.java:91) ~[cfg4j-core-4.4.1.jar:na]
	at org.cfg4j.provider.SimpleConfigurationProvider.getProperty(SimpleConfigurationProvider.java:76) ~[cfg4j-core-4.4.1.jar:na]
	at org.cfg4j.provider.BindInvocationHandler.invoke(BindInvocationHandler.java:61) ~[cfg4j-core-4.4.1.jar:na]
	at jdk.proxy2/jdk.proxy2.$Proxy54.listCarMessage(Unknown Source) ~[na:na]
	at by.miaskor.bot.service.handler.list.ListCarHandler$sendCars$2$1.invoke(ListCarHandler.kt:46) ~[main/:na]
	at by.miaskor.bot.service.handler.list.ListCarHandler$sendCars$2$1.invoke(ListCarHandler.kt:46) ~[main/:na]
	at by.miaskor.bot.service.MessageSender.sendMessage$lambda-4(MessageSender.kt:65) ~[main/:na]
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:113) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1816) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.MonoFlatMap$FlatMapInner.onNext(MonoFlatMap.java:249) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:120) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:74) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onNext(FluxHandleFuseable.java:191) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.onNext(FluxMapFuseable.java:295) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2398) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.request(FluxMapFuseable.java:354) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.request(FluxHandleFuseable.java:257) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.set(Operators.java:2194) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onSubscribe(Operators.java:2068) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onSubscribe(FluxHandleFuseable.java:161) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.onSubscribe(FluxMapFuseable.java:263) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.MonoJust.subscribe(MonoJust.java:55) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:157) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:127) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxHide$SuppressFuseableSubscriber.onNext(FluxHide.java:137) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onNext(FluxPeek.java:200) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxDefaultIfEmpty$DefaultIfEmptySubscriber.onNext(FluxDefaultIfEmpty.java:101) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:74) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2398) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.set(Operators.java:2194) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onSubscribe(Operators.java:2068) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.MonoJust.subscribe(MonoJust.java:55) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4400) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:203) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4385) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:200) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4385) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.subscribeNext(MonoIgnoreThen.java:263) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.MonoIgnoreThen.subscribe(MonoIgnoreThen.java:51) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4400) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.drain(FluxConcatMap.java:451) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.onNext(FluxConcatMap.java:243) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.runBackfused(FluxPublishOn.java:484) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.run(FluxPublishOn.java:521) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:84) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:37) ~[reactor-core-3.4.17.jar:3.4.17]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:833) ~[na:na]

2022-08-22 01:49:01.014 ERROR 412009 --- [nder-thread-354] reactor.core.publisher.Operators         : Operator called default onErrorDropped

java.lang.RuntimeException: java.io.InterruptedIOException: interrupted
	at com.pengrad.telegrambot.impl.TelegramBotClient.send(TelegramBotClient.java:74) ~[java-telegram-bot-api-6.0.0.jar:na]
	at com.pengrad.telegrambot.TelegramBot.execute(TelegramBot.java:44) ~[java-telegram-bot-api-6.0.0.jar:na]
	at by.miaskor.bot.service.extension.TelegramBotExtensionKt.sendMessage(TelegramBotExtension.kt:9) ~[main/:na]
	at by.miaskor.bot.service.MessageSender.sendMessage$lambda-4(MessageSender.kt:66) ~[main/:na]
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:113) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1816) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.MonoFlatMap$FlatMapInner.onNext(MonoFlatMap.java:249) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:120) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:74) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onNext(FluxHandleFuseable.java:191) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.onNext(FluxMapFuseable.java:295) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2398) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.request(FluxMapFuseable.java:354) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.request(FluxHandleFuseable.java:257) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.set(Operators.java:2194) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onSubscribe(Operators.java:2068) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onSubscribe(FluxHandleFuseable.java:161) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.onSubscribe(FluxMapFuseable.java:263) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.MonoJust.subscribe(MonoJust.java:55) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:157) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:127) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxHide$SuppressFuseableSubscriber.onNext(FluxHide.java:137) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onNext(FluxPeek.java:200) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxDefaultIfEmpty$DefaultIfEmptySubscriber.onNext(FluxDefaultIfEmpty.java:101) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:74) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2398) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.set(Operators.java:2194) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onSubscribe(Operators.java:2068) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.MonoJust.subscribe(MonoJust.java:55) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4400) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:203) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4385) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:200) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4385) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.subscribeNext(MonoIgnoreThen.java:263) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.MonoIgnoreThen.subscribe(MonoIgnoreThen.java:51) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4400) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.drain(FluxConcatMap.java:451) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.onNext(FluxConcatMap.java:243) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.runBackfused(FluxPublishOn.java:484) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.run(FluxPublishOn.java:521) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:84) ~[reactor-core-3.4.17.jar:3.4.17]
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:37) ~[reactor-core-3.4.17.jar:3.4.17]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:833) ~[na:na]
Caused by: java.io.InterruptedIOException: interrupted
	at okio.Timeout.throwIfReached(Timeout.java:146) ~[okio-1.17.2.jar:na]
	at okio.Okio$1.write(Okio.java:76) ~[okio-1.17.2.jar:na]
	at okio.AsyncTimeout$1.write(AsyncTimeout.java:180) ~[okio-1.17.2.jar:na]
	at okio.RealBufferedSink.flush(RealBufferedSink.java:224) ~[okio-1.17.2.jar:na]
	at okhttp3.internal.http2.Http2Writer.flush(Http2Writer.java:121) ~[okhttp-3.14.9.jar:na]
	at okhttp3.internal.http2.Http2Connection.flush(Http2Connection.java:439) ~[okhttp-3.14.9.jar:na]
	at okhttp3.internal.http2.Http2Stream$FramingSink.close(Http2Stream.java:631) ~[okhttp-3.14.9.jar:na]
	at okio.ForwardingSink.close(ForwardingSink.java:47) ~[okio-1.17.2.jar:na]
	at okhttp3.internal.connection.Exchange$RequestBodySink.close(Exchange.java:253) ~[okhttp-3.14.9.jar:na]
	at okio.RealBufferedSink.close(RealBufferedSink.java:248) ~[okio-1.17.2.jar:na]
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:70) ~[okhttp-3.14.9.jar:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.9.jar:na]
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:43) ~[okhttp-3.14.9.jar:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.9.jar:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[okhttp-3.14.9.jar:na]
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94) ~[okhttp-3.14.9.jar:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.9.jar:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[okhttp-3.14.9.jar:na]
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) ~[okhttp-3.14.9.jar:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.9.jar:na]
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:88) ~[okhttp-3.14.9.jar:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.9.jar:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[okhttp-3.14.9.jar:na]
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:229) ~[okhttp-3.14.9.jar:na]
	at okhttp3.RealCall.execute(RealCall.java:81) ~[okhttp-3.14.9.jar:na]
	at com.pengrad.telegrambot.impl.TelegramBotClient.send(TelegramBotClient.java:71) ~[java-telegram-bot-api-6.0.0.jar:na]
	... 50 common frames omitted
```

--- a/bamper-integration/bamper-integration/src/main/kotlin/by/miaskor/service/ExcelGenerator.kt
+++ b/bamper-integration/bamper-integration/src/main/kotlin/by/miaskor/service/ExcelGenerator.kt
@@ -35,7 +35,6 @@ class ExcelGenerator(
 
   private fun fillValues(row: Int = 1, worksheet: Worksheet, autoPartResponse: AutoPartResponse) {
     worksheet.apply {
-      value(row, 0, 123)
       value(row, 1, autoPartResponse.brand)
       value(row, 2, autoPartResponse.model)
       value(row, 4, autoPartResponse.year)

--- a/bot/src/main/kotlin/by/miaskor/bot/BotApplication.kt
+++ b/bot/src/main/kotlin/by/miaskor/bot/BotApplication.kt
@@ -1,12 +1,16 @@
 package by.miaskor.bot
 
+import by.miaskor.common.DelegatePropertyConfiguration
 import by.miaskor.common.PropertyConfiguration
 import org.springframework.boot.autoconfigure.SpringBootApplication
 import org.springframework.boot.runApplication
 import org.springframework.context.annotation.Import
 import org.springframework.context.annotation.PropertySource
 
-@Import(PropertyConfiguration::class)
+@Import(
+  PropertyConfiguration::class,
+  DelegatePropertyConfiguration::class
+)
 @PropertySource("classpath:bot-application.properties")
 @SpringBootApplication
 open class BotApplication

--- a/bot/src/main/kotlin/by/miaskor/bot/configuration/BotConfiguration.kt
+++ b/bot/src/main/kotlin/by/miaskor/bot/configuration/BotConfiguration.kt
@@ -24,8 +24,7 @@ open class BotConfiguration(
   open fun bot(): Bot {
     return Bot(
       telegramBot(),
-      cacheConfiguration.telegramClientCache(),
-      settingsConfiguration.executorSettings()
+      cacheConfiguration.telegramClientCache()
     )
   }
 }

--- a/bot/src/main/kotlin/by/miaskor/bot/configuration/Cf4jConfiguration.kt
+++ b/bot/src/main/kotlin/by/miaskor/bot/configuration/Cf4jConfiguration.kt
@@ -1,9 +1,9 @@
 package by.miaskor.bot.configuration
 
+import by.miaskor.bot.service.cf4j.DefaultConfigurationProvider
 import by.miaskor.bot.service.cf4j.LinkPropertiesProvider
 import by.miaskor.common.PropertyConfiguration
 import org.cfg4j.provider.ConfigurationProvider
-import org.cfg4j.provider.ConfigurationProviderBuilder
 import org.cfg4j.source.context.propertiesprovider.JsonBasedPropertiesProvider
 import org.cfg4j.source.context.propertiesprovider.PropertiesProviderSelector
 import org.cfg4j.source.context.propertiesprovider.PropertyBasedPropertiesProvider
@@ -27,8 +27,6 @@ open class Cf4jConfiguration(
 
   @Bean
   open fun configurationProvider(): ConfigurationProvider {
-    return ConfigurationProviderBuilder()
-      .withConfigurationSource(classPathConfigurationSource())
-      .build()
+    return DefaultConfigurationProvider(classPathConfigurationSource())
   }
 }

--- a/bot/src/main/kotlin/by/miaskor/bot/service/cf4j/DefaultConfigurationProvider.kt
+++ b/bot/src/main/kotlin/by/miaskor/bot/service/cf4j/DefaultConfigurationProvider.kt
@@ -1,0 +1,99 @@
+package by.miaskor.bot.service.cf4j
+
+import com.github.drapostolos.typeparser.NoSuchRegisteredParserException
+import com.github.drapostolos.typeparser.TypeParser
+import com.github.drapostolos.typeparser.TypeParserException
+import org.cfg4j.provider.ConfigurationProvider
+import org.cfg4j.provider.GenericTypeInterface
+import org.cfg4j.source.ConfigurationSource
+import org.cfg4j.source.context.environment.DefaultEnvironment
+import org.cfg4j.source.context.environment.Environment
+import org.cfg4j.source.context.environment.MissingEnvironmentException
+import java.lang.reflect.Proxy
+import java.util.*
+import java.util.concurrent.ConcurrentHashMap
+
+class DefaultConfigurationProvider(
+  private val configurationSource: ConfigurationSource,
+  private val environment: Environment = DefaultEnvironment(),
+) : ConfigurationProvider {
+  override fun allConfigurationAsProperties(): Properties {
+    return try {
+      configurationSource.getConfiguration(environment)
+    } catch (e: IllegalStateException) {
+      throw IllegalStateException("Couldn't fetch configuration from configuration source", e)
+    } catch (e: MissingEnvironmentException) {
+      throw IllegalStateException("Couldn't fetch configuration from configuration source", e)
+    }
+  }
+
+  override fun <T : Any> getProperty(key: String, type: Class<T>): T {
+    val propertyStr = getProperty(key)
+
+    return try {
+      val parser = TypeParser.newBuilder().build()
+      parser.parse(propertyStr, type)
+    } catch (e: TypeParserException) {
+      throw IllegalArgumentException("Unable to cast value \'$propertyStr\' to $type", e)
+    } catch (e: NoSuchRegisteredParserException) {
+      throw IllegalArgumentException("Unable to cast value \'$propertyStr\' to $type", e)
+    }
+  }
+
+  override fun <T : Any> getProperty(key: String, genericType: GenericTypeInterface): T {
+    val propertyStr = getProperty(key)
+
+    return try {
+      val parser =
+        TypeParser.newBuilder().build()
+      parser.parseType(propertyStr, genericType.type) as T
+    } catch (e: TypeParserException) {
+      throw IllegalArgumentException("Unable to cast value \'$propertyStr\' to $genericType", e)
+    } catch (e: NoSuchRegisteredParserException) {
+      throw IllegalArgumentException("Unable to cast value \'$propertyStr\' to $genericType", e)
+    }
+  }
+
+  override fun <T : Any> bind(prefix: String, type: Class<T>): T {
+    val allProperties = getAllProperties(prefix, type)
+    return bindObject(allProperties, type)
+  }
+
+  private fun <T : Any> bindObject(properties: Map<String, Any>, type: Class<T>): T {
+    return Proxy.newProxyInstance(
+      type.classLoader,
+      arrayOf<Class<*>>(type)
+    ) { proxy, method, methodArgs ->
+      val any = properties[method.name]
+      if (any is Map<*, *>) {
+        bindObject(any as Map<String, Any>, method.returnType)
+      } else {
+        any
+      }
+    } as T
+  }
+
+  private fun <T : Any> getAllProperties(prefix: String, type: Class<T>): Map<String, Any> {
+    val map = ConcurrentHashMap<String, Any>()
+
+    type.declaredMethods.forEach {
+      if (it.returnType.isInterface && it.returnType.simpleName.contains("Settings")) {
+        map[it.name] = getAllProperties("$prefix.${it.name}", it.returnType)
+      } else {
+        map[it.name] = getProperty("$prefix.${it.name}", it.returnType)
+      }
+    }
+
+    return map
+  }
+
+  private fun getProperty(key: String): String {
+    return try {
+      val property = configurationSource.getConfiguration(environment)[key]
+        ?: throw NoSuchElementException("No configuration with key: $key")
+      property.toString()
+    } catch (e: IllegalStateException) {
+      throw IllegalStateException("Couldn't fetch configuration from configuration source for key: $key", e)
+    }
+  }
+}

--- a/bot/src/main/kotlin/by/miaskor/bot/service/handler/command/BackCommandHandler.kt
+++ b/bot/src/main/kotlin/by/miaskor/bot/service/handler/command/BackCommandHandler.kt
@@ -1,6 +1,7 @@
 package by.miaskor.bot.service.handler.command
 
 import by.miaskor.bot.configuration.settings.MessageSettings
+import by.miaskor.bot.domain.BotState.BAMPER_MENU
 import by.miaskor.bot.domain.BotState.CHOOSING_STORE_HOUSE
 import by.miaskor.bot.domain.BotState.EMPLOYEES_MENU
 import by.miaskor.bot.domain.BotState.FINDING_AUTO_PART
@@ -54,6 +55,7 @@ class BackCommandHandler(
             .format(telegramClient.currentStoreHouseName())
 
           FINDING_AUTO_PART -> messageSettings.findAutoPartMenuMessage()
+          BAMPER_MENU -> messageSettings.bamperMenuMessage()
           else -> "Something bad happened"
         }
         telegramBot.sendMessageWithKeyboard(update.chatId, sendMessage, keyboard)

--- a/bot/src/main/kotlin/by/miaskor/bot/service/handler/state/CreatingAutoPartHandler.kt
+++ b/bot/src/main/kotlin/by/miaskor/bot/service/handler/state/CreatingAutoPartHandler.kt
@@ -47,13 +47,13 @@ class CreatingAutoPartHandler(
   ): Mono<Unit> {
     return Mono.just(update.chatId)
       .flatMap(telegramClientCache::getTelegramClient)
-      .flatMap { completeCreatingCar(update, it, stepBuilder as AutoPartBuilder) }
+      .flatMap { completeCreatingAutoPart(update, it, stepBuilder as AutoPartBuilder) }
   }
 
-  private fun completeCreatingCar(
+  private fun completeCreatingAutoPart(
     update: Update,
     telegramClient: TelegramClient,
-    autoPartBuilder: AutoPartBuilder
+    autoPartBuilder: AutoPartBuilder,
   ): Mono<Unit> {
     return Mono.just(update.chatId)
       .changeBotState(update::chatId, telegramClient.previousBotStates.pollLast())

--- a/bot/src/main/kotlin/by/miaskor/bot/service/handler/state/CreatingCarHandler.kt
+++ b/bot/src/main/kotlin/by/miaskor/bot/service/handler/state/CreatingCarHandler.kt
@@ -58,7 +58,7 @@ class CreatingCarHandler(
     return Mono.just(update.chatId)
       .changeBotState(update::chatId, telegramClient.previousBotStates.pollLast())
       .flatMap { carConnector.create(carStepBuilder.build(telegramClient.currentStoreHouseId())) }
-      .map { carId ->
+      .flatMap { carId ->
         sendMessageWithKeyboard<Long>(update.chatId, MessageSettings::completeCreatingCarMessage, carId.toString())
       }
       .then(Mono.empty())

--- a/bot/src/main/kotlin/by/miaskor/bot/service/handler/state/StepHandler.kt
+++ b/bot/src/main/kotlin/by/miaskor/bot/service/handler/state/StepHandler.kt
@@ -3,8 +3,8 @@ package by.miaskor.bot.service.handler.state
 import by.miaskor.bot.configuration.settings.KeyboardSettings
 import by.miaskor.bot.domain.AbstractStep
 import by.miaskor.bot.domain.AbstractStepBuilder
-import by.miaskor.bot.domain.Command
 import by.miaskor.bot.domain.Command.NEXT_STEP
+import by.miaskor.bot.domain.Command.PREVIOUS_STEP
 import by.miaskor.bot.service.LanguageSettingsResolver.resolveLanguage
 import by.miaskor.bot.service.cache.Cache
 import by.miaskor.bot.service.extension.chatId
@@ -36,24 +36,30 @@ abstract class StepHandler<T>(
           cache.populate(update.chatId, nextStep)
           sendMessage(nextStep, update)
             .filter { currentStep.next().isFinalStep() }
-            .flatMap { completeStep(update, stepBuilder) }
+            .flatMap {
+              completeStep(update, stepBuilder)
+                .doOnTerminate { cache.evict(update.chatId) }
+            }
         } else if (NEXT_STEP isCommand update.text && !currentStep.isStepNotMandatory()) {
           sendMessage(stepBuilder, update)
-        } else if (Command.PREVIOUS_STEP isCommand update.text) {
+        } else if (PREVIOUS_STEP isCommand update.text) {
           val previousStep = stepBuilder.previousStep()
           cache.populate(update.chatId, previousStep)
           sendMessage(previousStep, update)
         } else {
           processingStepService.process(update, currentStep, stepBuilder)
             .filter { it && stepBuilder.currentStep().isFinalStep() }
-            .flatMap { completeStep(update, stepBuilder) }
+            .flatMap {
+              completeStep(update, stepBuilder)
+                .doOnTerminate { cache.evict(update.chatId) }
+            }
         }
       }
   }
 
   protected fun sendMessage(
     stepBuilder: AbstractStepBuilder<T>,
-    update: Update
+    update: Update,
   ): Mono<Unit> {
     return Mono.just(update.chatId)
       .resolveLanguage(KeyboardSettings::class)
@@ -66,6 +72,6 @@ abstract class StepHandler<T>(
 
   protected abstract fun completeStep(
     update: Update,
-    stepBuilder: AbstractStepBuilder<T>
+    stepBuilder: AbstractStepBuilder<T>,
   ): Mono<Unit>
 }

--- a/bot/src/main/kotlin/by/miaskor/bot/service/step/car/CarBuilderFieldEnricher.kt
+++ b/bot/src/main/kotlin/by/miaskor/bot/service/step/car/CarBuilderFieldEnricher.kt
@@ -8,6 +8,7 @@ import by.miaskor.bot.domain.CreatingCarStep.BRAND_NAME
 import by.miaskor.bot.domain.CreatingCarStep.ENGINE_CAPACITY
 import by.miaskor.bot.domain.CreatingCarStep.ENGINE_TYPE
 import by.miaskor.bot.domain.CreatingCarStep.FUEL_TYPE
+import by.miaskor.bot.domain.CreatingCarStep.MODEL
 import by.miaskor.bot.domain.CreatingCarStep.TRANSMISSION
 import by.miaskor.bot.domain.CreatingCarStep.YEAR
 import by.miaskor.bot.service.extension.text
@@ -24,6 +25,7 @@ class CarBuilderFieldEnricher : StepBuilderFieldEnricher<CreatingCarStep> {
     val carBuilder = stepBuilder as CarBuilder
     when (step) {
       BRAND_NAME -> carBuilder.brandName(update.text)
+      MODEL -> carBuilder.model(update.text)
       YEAR -> carBuilder.year(update.text)
       BODY -> carBuilder.body(update.text)
       TRANSMISSION -> carBuilder.transmission(update.text)

--- a/bot/src/main/kotlin/by/miaskor/bot/telegram/Bot.kt
+++ b/bot/src/main/kotlin/by/miaskor/bot/telegram/Bot.kt
@@ -5,6 +5,7 @@ import by.miaskor.bot.service.CommandResolver.processCommand
 import by.miaskor.bot.service.cache.TelegramClientCache
 import by.miaskor.bot.service.extension.chatId
 import by.miaskor.bot.service.extension.info
+import by.miaskor.common.BindingProperty
 import com.pengrad.telegrambot.TelegramBot
 import com.pengrad.telegrambot.UpdatesListener.CONFIRMED_UPDATES_ALL
 import com.pengrad.telegrambot.model.Update
@@ -16,8 +17,8 @@ import reactor.core.scheduler.Schedulers
 class Bot(
   telegramBot: TelegramBot,
   private val telegramClientCache: TelegramClientCache,
-  private val executorSettings: ExecutorSettings,
 ) {
+  private val executorSettings: ExecutorSettings by BindingProperty("bot.executor")
 
   init {
     telegramBot.setUpdatesListener { updates ->

--- a/common/build.gradle
+++ b/common/build.gradle
@@ -1,2 +1,0 @@
-dependencies {
-}

--- a/common/build.gradle.kts
+++ b/common/build.gradle.kts
@@ -1,0 +1,4 @@
+dependencies {
+
+  implementation(lib("io.projectreactor:reactor-core"))
+}

--- a/common/src/main/kotlin/by/miaskor/common/BindingProperty.kt
+++ b/common/src/main/kotlin/by/miaskor/common/BindingProperty.kt
@@ -1,0 +1,12 @@
+package by.miaskor.common
+
+import kotlin.reflect.KClass
+import kotlin.reflect.KProperty
+
+class BindingProperty<V : Any>(
+  private val bindPath: String,
+) : Property<V>() {
+  override fun propertyProvider(thisRef: Any, property: KProperty<*>): V {
+    return configurationProvider.bind(bindPath, (property.returnType.classifier as KClass<V>).javaObjectType) as V
+  }
+}

--- a/common/src/main/kotlin/by/miaskor/common/DefaultProperty.kt
+++ b/common/src/main/kotlin/by/miaskor/common/DefaultProperty.kt
@@ -1,0 +1,12 @@
+package by.miaskor.common
+
+import kotlin.reflect.KProperty
+import kotlin.reflect.jvm.javaField
+
+class DefaultProperty<V : Any>(
+  private val propertyPath: String,
+) : Property<V>() {
+  override fun propertyProvider(thisRef: Any, property: KProperty<*>): V {
+    return configurationProvider.getProperty(propertyPath, property.javaField?.type) as V
+  }
+}

--- a/common/src/main/kotlin/by/miaskor/common/DelegatePropertyConfiguration.kt
+++ b/common/src/main/kotlin/by/miaskor/common/DelegatePropertyConfiguration.kt
@@ -1,0 +1,15 @@
+package by.miaskor.common
+
+import org.cfg4j.provider.ConfigurationProvider
+import org.springframework.context.annotation.Configuration
+import javax.annotation.PostConstruct
+
+@Configuration
+open class DelegatePropertyConfiguration(
+  private val configurationProvider: ConfigurationProvider,
+) {
+  @PostConstruct
+  fun init() {
+    Property.configurationProvider = configurationProvider
+  }
+}

--- a/common/src/main/kotlin/by/miaskor/common/Property.kt
+++ b/common/src/main/kotlin/by/miaskor/common/Property.kt
@@ -1,0 +1,28 @@
+package by.miaskor.common
+
+import org.cfg4j.provider.ConfigurationProvider
+import java.util.concurrent.atomic.AtomicReference
+import kotlin.properties.ReadOnlyProperty
+import kotlin.reflect.KProperty
+
+abstract class Property<V : Any> : ReadOnlyProperty<Any, V> {
+
+  private val backField = AtomicReference<V>()
+
+  override fun getValue(thisRef: Any, property: KProperty<*>): V {
+    return if (backField.get() == null) {
+      backField.updateAndGet { propertyProvider(thisRef, property) }
+    } else {
+      backField.get()
+    }
+  }
+
+  protected abstract fun propertyProvider(thisRef: Any, property: KProperty<*>): V
+
+  companion object {
+    lateinit var configurationProvider: ConfigurationProvider
+  }
+}
+
+
+

--- a/domain/domain/src/main/kotlin/by/miaskor/domain/DomainApplication.kt
+++ b/domain/domain/src/main/kotlin/by/miaskor/domain/DomainApplication.kt
@@ -1,6 +1,7 @@
 package by.miaskor.domain
 
 import by.miaskor.cloud.drive.configuration.CloudDriveConfiguration
+import by.miaskor.common.DelegatePropertyConfiguration
 import by.miaskor.common.PropertyConfiguration
 import org.springframework.boot.autoconfigure.SpringBootApplication
 import org.springframework.boot.runApplication
@@ -9,7 +10,8 @@ import org.springframework.context.annotation.PropertySource
 
 @Import(
   CloudDriveConfiguration::class,
-  PropertyConfiguration::class
+  PropertyConfiguration::class,
+  DelegatePropertyConfiguration::class,
 )
 @PropertySource("classpath:domain-application.properties")
 @SpringBootApplication

--- a/domain/domain/src/main/kotlin/by/miaskor/domain/configuration/ServiceConfiguration.kt
+++ b/domain/domain/src/main/kotlin/by/miaskor/domain/configuration/ServiceConfiguration.kt
@@ -74,13 +74,7 @@ open class ServiceConfiguration(
 
   @Bean
   open fun telegramConnector(): TelegramConnector {
-    val getPhotoPathUrl = confProvider.getProperty("bot.getPhotoPathUri", String::class.java)
-    val getPhotoUrl = confProvider.getProperty("bot.getPhotoUri", String::class.java)
-    return TelegramConnector(
-      connectorConfiguration.telegramWebClient(),
-      getPhotoPathUrl,
-      getPhotoUrl
-    )
+    return TelegramConnector(connectorConfiguration.telegramWebClient())
   }
 
   @Bean

--- a/domain/domain/src/main/kotlin/by/miaskor/domain/service/StoreHouseService.kt
+++ b/domain/domain/src/main/kotlin/by/miaskor/domain/service/StoreHouseService.kt
@@ -7,8 +7,6 @@ import by.miaskor.domain.service.mapper.StoreHouseMapper
 import by.miaskor.domain.tables.pojos.StoreHouse
 import reactor.core.publisher.Flux
 import reactor.core.publisher.Mono
-import reactor.kotlin.core.util.function.component1
-import reactor.kotlin.core.util.function.component2
 
 class StoreHouseService(
   private val storeHouseRepository: StoreHouseRepository,
@@ -34,14 +32,18 @@ class StoreHouseService(
   private fun getWorkerStoreHouses(workerStoreHouses: List<WorkerStoreHouseVO>): Flux<StoreHouseDto> {
     return Mono.fromSupplier { workerStoreHouses.map { it.storeHouseId } }
       .flatMap { workerStoreHouseIds -> storeHouseRepository.findAllByIds(workerStoreHouseIds) }
-      .flatMapIterable { it }
-      .zipWithIterable(workerStoreHouses)
-      .map { (storeHouse, workerStoreHouseVO) ->
-        StoreHouseDto(
-          id = storeHouse.id ?: -1,
-          chatId = storeHouse.telegramChatId ?: -1,
-          name = storeHouse.storeHouseName ?: "",
-          modifiable = workerStoreHouseVO.privilege == "M"
+      .flatMapMany { storeHouses ->
+        Flux.fromIterable(
+          storeHouses.map { storeHouse ->
+            val workerStoreHouseVO = workerStoreHouses
+              .find { workerStoreHouse -> workerStoreHouse.storeHouseId == storeHouse.id }
+            StoreHouseDto(
+              id = storeHouse.id ?: -1,
+              chatId = storeHouse.telegramChatId ?: -1,
+              name = storeHouse.storeHouseName ?: "",
+              modifiable = workerStoreHouseVO?.privilege == "M"
+            )
+          }
         )
       }
   }

--- a/domain/domain/src/main/kotlin/by/miaskor/domain/service/telegram/TelegramConnector.kt
+++ b/domain/domain/src/main/kotlin/by/miaskor/domain/service/telegram/TelegramConnector.kt
@@ -1,5 +1,6 @@
 package by.miaskor.domain.service.telegram
 
+import by.miaskor.common.DefaultProperty
 import by.miaskor.domain.service.telegram.domain.PhotoPathResponse
 import org.springframework.core.io.buffer.DataBuffer
 import org.springframework.http.MediaType
@@ -9,9 +10,9 @@ import reactor.core.publisher.Mono
 
 class TelegramConnector(
   private val webClient: WebClient,
-  private val getPhotoPathUri: String,
-  private val getPhotoUri: String,
 ) {
+  private val getPhotoPathUri: String by DefaultProperty("bot.getPhotoPathUri")
+  private val getPhotoUri: String by DefaultProperty("bot.getPhotoUri")
 
   fun getPhotoPath(photoId: String): Mono<String> {
     return webClient

